File: mission-to-mars-solid/src/main/java/mars/DirectionHandlerFactory.java
package mars;

import java.util.EnumMap;
import java.util.function.Function;

public class DirectionHandlerFactory implements Function<Direction, Mouvement2D> {

	private final EnumMap<Direction, Mouvement2D> handlers = new EnumMap<>(Direction.class);

	public DirectionHandlerFactory() {
		handlers.put(Direction.NORD, new MouvementNord2D());
		handlers.put(Direction.SUD, new MouvementSud2D());
		handlers.put(Direction.OUEST, new MouvementOuest2D());
		handlers.put(Direction.EST, new MouvementEst2D());
	}

	@Override
	public Mouvement2D apply(Direction direction) {
		return handlers.get(direction);
	}

}

File: python/mission-to-mars-solid/src/mars/DirectionHandlerFactory.py
from src.mars.Direction import Direction
from src.mars.Mouvement2D import Mouvement2D
from src.mars.MouvementEst2D import MouvementEst2D
from src.mars.MouvementNord2D import MouvementNord2D
from src.mars.MouvementOuest2D import MouvementOuest2D
from src.mars.MouvementSud2D import MouvementSud2D


class DirectionHandlerFactory:
    def __init__(self):
        self.handlers = {
            Direction.NORD: MouvementNord2D(),
            Direction.SUD: MouvementSud2D(),
            Direction.EST: MouvementEst2D(),
            Direction.OUEST: MouvementOuest2D(),
        }

    def apply(self, direction: Direction) -> Mouvement2D:
        return self.handlers.get(direction)
